File: haskell/perfect-numbers/src/PerfectNumbers.hs
module PerfectNumbers (classify, Classification(..)) where

data Classification = Deficient | Perfect | Abundant deriving (Eq, Show)

classify :: Int -> Maybe Classification
classify x
  | x < 1 = Nothing
  | sum (factorsOf x) > x = Just Abundant
  | sum (factorsOf x) == x = Just Perfect
  | sum (factorsOf x) < x = Just Deficient
  | otherwise = Nothing

isFactorOf :: Int -> Int -> Bool
isFactorOf factor num = num `rem` factor == 0

factorsOf :: Int -> [Int]
factorsOf x = filter (`isFactorOf` x) [1..x-1]

File: javascript/perfect-numbers/perfect-numbers.js
export const classify = (num) => {

  if (num < 1) throw new Error('Classification is only possible for natural numbers.');

  let classification = sumOfArray(factorise(num));

  if (classification > num) {
    return 'abundant';
  } else if (classification < num) {
    return 'deficient';
  } else {
    return 'perfect';
  }
}

export const sumOfArray = (arr) => arr.reduce( (acc, cur) => { return acc + cur }, 0)

export const factorise = (num) => {
  let out = [];
  
  for (let i = 0; i < num; i++) { if (num % i === 0) out.push(i) }

  return out;
}
